File: problems/623-add-one-row-to-tree/dfs-v2.js
/**
 * DFS：带返回值
 * 
 * 时间：`O(N)`, 80ms
 */
var addOneRow = function (root, v, d) {
  
  function dfs (node, depth, isLeft) {
    // d层时，返回新结点，给父结点来链接
    if (depth === d) {
      const newNode = new TreeNode(v)
      if (isLeft) {
        newNode.left = node
      } else {
        newNode.right = node
      }
      return newNode
    }
    
    if (!node) return node

    // 重新链接
    node.left = dfs(node.left, depth + 1, true)
    node.right = dfs(node.right, depth + 1, false)

    // 返回当前结点
    return node
  }

  return dfs(root, 1, true) // 统一了特殊情况
};

File: biweekly-contest/biweekly-contest-17/3-sum-of-nodes-with-even-valued-grandparent/dfs-v2.cpp
/**
 * DFS：记录当前路径
 */

class Solution {
public:
    int sumEvenGrandparent(TreeNode* root) {
        dfs(root);
        return res;
    }

    void dfs(TreeNode* node) {
        if (!node) return;

        // 探索
        path.push_back(node->val);
        if (path.size() >= 3 && path[path.size() - 3] % 2 == 0) {
            res += node->val;
        }
        dfs(node->left);
        dfs(node->right);

        // 回溯
        path.pop_back();
    }

private:
    int res;
    vector<int> path;
};